src/puppetlabs/pcp/client.clj

```diff
@@ -203,8 +203,10 @@
                                     (log/debug (i18n/trs "WebSocket closed {0} {1}" (str code) message))
                                     (reset! associate-response (promise))
                                     (let [{:keys [should-stop websocket-connection]} client]
+                                      ;; Ensure disconnect state is immediately registered.
+                                      (reset! websocket-connection (future true))
```

Comment:
This still isn't quite equivalent: before it would've immediately gone to connecting, here it immediately goes to disconnected (so `wait-for-connection` returns immediately).